File: samples/mdi/mainwindow.nim
import ../../vendor/seaqt/seaqt/[
  qmainwindow, qmdiarea, qmdisubwindow, qpushbutton, qtoolbar, qtabbar, qaction
]

import ./mdichild

type
  MainWindow* = ref object
    window*: QMainWindow
    mdiArea: QMdiArea
    addButton: QPushButton
    toggleButton: QPushButton
    mainToolBar: QToolBar
    windowCounter: int

proc addMdiWindow(self: MainWindow) =
  # # Create a new MDI child window
  let child = newMdiChild()
  self.windowCounter += 1
  child.setWindowTitle("MDI Window " & $self.windowCounter)
  
  # # Add the child to the MDI area
  let subWindow = self.mdiArea.addSubWindow(child.widget)
  # Reset the owner flag since addSubWindow takes ownership
  child.widget.owned = false
  subWindow.setMinimumSize(200, 200)
  subWindow.show()

proc toggleViewMode(self: MainWindow) =
  # Toggle between tabbed and windowed mode
  if self.mdiArea.viewMode() == 0: # SubWindowView = 0
    self.mdiArea.setViewMode(1)    # TabbedView = 1
    self.toggleButton.setText("Switch to Windowed")
  else:
    self.mdiArea.setViewMode(0)    # SubWindowView = 0
    self.toggleButton.setText("Switch to Tabbed")

proc setupMdiArea(self: MainWindow) =
  # Set up the MDI area
  self.mdiArea = QMdiArea.create(self.window)
  self.mdiArea.setHorizontalScrollBarPolicy(0) # ScrollBarAsNeeded = 0
  self.mdiArea.setVerticalScrollBarPolicy(0)   # ScrollBarAsNeeded = 0
  self.mdiArea.setViewMode(0) # SubWindowView = 0, start in windowed mode

  # Enable tab close buttons
  self.mdiArea.setTabsClosable(true)
  
  # ================================================
  # TODO: connect the tabCloseRequested signal to handle tab closing
  # ================================================

  # Add an initial MDI window
  self.addMdiWindow()

proc createToolBar(self: MainWindow) =
  # Create toolbar with buttons
  # self.mainToolBar = self.window.addToolBar("Main")
  self.mainToolBar = self.window.addToolBar("Main")
  # self.mainToolBar = self.addToolBar("Main")

  # Add button
  self.addButton = QPushButton.create("Add Window", self.window)
  self.addButton.onPressed(proc() = self.addMdiWindow())

  discard self.mainToolBar.addWidget(self.addButton)
  # Reset the owner flag since addWidget takes ownership
  self.addButton.owned = false

  # Toggle button
  self.toggleButton = QPushButton.create("Toggle View Mode", self.window)
  self.toggleButton.onPressed(proc() = self.toggleViewMode())

  discard self.mainToolBar.addWidget(self.toggleButton)
  # Reset the owner flag since addWidget takes ownership
  self.toggleButton.owned = false

proc newMainWindow*(): MainWindow =
  result = MainWindow(
    window: QMainWindow.create(),
    windowCounter: 0
  )
  
  # Set up the MDI area
  result.setupMdiArea()
  
  # Create toolbar with buttons
  result.createToolBar()
  
  # Set central widget
  result.window.setCentralWidget(result.mdiArea)
  
  # Set initial size
  result.window.resize(800, 600)

proc show*(self: MainWindow) =
  self.window.show()

proc setWindowTitle*(self: MainWindow, title: string) =
  self.window.setWindowTitle(title) 

File: reference/mdi/mainwindow.cpp
#include "mainwindow.h"
#include "mdichild.h"
#include <QMdiSubWindow>
#include <QToolBar>
#include <QVBoxLayout>
#include <QLabel>
#include <QTabBar>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), windowCounter(0)
{
    // Set up the MDI area
    setupMdiArea();
    
    // Create toolbar with buttons
    createToolBar();
    
    // Set central widget
    setCentralWidget(mdiArea);
    
    // Set initial size
    resize(800, 600);
}

MainWindow::~MainWindow()
{
}

void MainWindow::setupMdiArea()
{
    mdiArea = new QMdiArea(this);
    mdiArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    mdiArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    mdiArea->setViewMode(QMdiArea::SubWindowView); // Start in windowed mode
    
    // Enable tab close buttons
    mdiArea->setTabsClosable(true);
    
    // Connect the tabCloseRequested signal to handle tab closing
    connect(mdiArea, &QMdiArea::subWindowActivated, this, &MainWindow::updateTabCloseButtons);
    
    // Add an initial MDI window
    addMdiWindow();
}

void MainWindow::createToolBar()
{
    mainToolBar = addToolBar(tr("Main"));
    
    // Add button
    addButton = new QPushButton(tr("Add Window"), this);
    connect(addButton, &QPushButton::clicked, this, &MainWindow::addMdiWindow);
    mainToolBar->addWidget(addButton);
    
    // Toggle button
    toggleButton = new QPushButton(tr("Toggle View Mode"), this);
    connect(toggleButton, &QPushButton::clicked, this, &MainWindow::toggleViewMode);
    mainToolBar->addWidget(toggleButton);
}

void MainWindow::addMdiWindow()
{
    // Create a new MDI child window
    MdiChild *child = new MdiChild;
    windowCounter++;
    child->setWindowTitle(tr("MDI Window %1").arg(windowCounter));
    
    // Add the child to the MDI area
    QMdiSubWindow *subWindow = mdiArea->addSubWindow(child);
    subWindow->setMinimumSize(200, 200);
    subWindow->show();
    
    // Connect the close event of the subwindow
    connect(subWindow, &QMdiSubWindow::windowStateChanged, this, &MainWindow::updateTabCloseButtons);
}

void MainWindow::toggleViewMode()
{
    // Toggle between tabbed and windowed mode
    if (mdiArea->viewMode() == QMdiArea::SubWindowView) {
        mdiArea->setViewMode(QMdiArea::TabbedView);
        toggleButton->setText(tr("Switch to Windowed"));
        
        // Make sure tab close buttons are visible
        updateTabCloseButtons();
    } else {
        mdiArea->setViewMode(QMdiArea::SubWindowView);
        toggleButton->setText(tr("Switch to Tabbed"));
    }
}

void MainWindow::updateTabCloseButtons()
{
    // This function ensures that tab close buttons are visible
    // It needs to be called when switching to tabbed mode and when windows change
    if (mdiArea->viewMode() == QMdiArea::TabbedView) {
        // Find the internal QTabBar
        QTabBar* tabBar = mdiArea->findChild<QTabBar*>();
        if (tabBar) {
            tabBar->setTabsClosable(true);
            
            // Disconnect any existing connections to avoid duplicates
            disconnect(tabBar, &QTabBar::tabCloseRequested, nullptr, nullptr);
            
            // Connect the tabCloseRequested signal
            connect(tabBar, &QTabBar::tabCloseRequested, [this](int index) {
                // Get the list of subwindows
                QList<QMdiSubWindow*> windows = mdiArea->subWindowList();
                
                // Make sure the index is valid
                if (index >= 0 && index < windows.size()) {
                    // Close the window at the specified index
                    windows.at(index)->close();
                }
            });
        }
    }
} 